### PR TITLE
Avoid overflowing PROCStack in BuildWaitGraphForSourceNode

### PR DESCRIPTION
**Pull request by marcocitus**
Currently, we may add the same process to the stack of remaining processes multiple times during the depth-first search in `BuildWaitGraphForSourceNode`, because we only inspect the visited status after a process has been added to the array. This means the size of the array (`MaxBackends`) may not be sufficient. 

After this change, the check is performed upfront to avoid adding the same process multiple times.

### CODE REVIEW

File: src/backend/distributed/transaction/lock_graph.c [hunk 1]
```diff
@@ -706,6 +687,23 @@ AllocWaitEdge(WaitGraph *waitGraph)
...
+ * in the depth-first search, unless it was already added.
+ */
+static void
+AddProcToVisit(PROCStack *remaining, PGPROC *proc)
+{
+	if (remaining->procAdded[proc->pgprocno])
+	{
+		return;
+	}
+
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/lock_graph.c]: I'd add an `Assert (remaining->procCount < MaxBackends)`

File: src/backend/distributed/transaction/lock_graph.c [hunk 2]
```diff
@@ -453,33 +453,14 @@ BuildWaitGraphForSourceNode(int sourceNodeId)
...
-		 * case of a local deadlock.
-		 */
-		if (visitedProcs[waitingProc->pgprocno])
-		{
-			continue;
-		}
-
-		visitedProcs[waitingProc->pgprocno] = true;
-
-		/* only blocked processes result in wait edges */
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/lock_graph.c]: Why are we removing this check? That seems not related to me?

Comment (marcocitus) [inline src/backend/distributed/transaction/lock_graph.c]: It's fairly redundant, but we can keep it.

File: src/backend/distributed/transaction/lock_graph.c (hunk 1 above)
Comment (marcocitus) [inline src/backend/distributed/transaction/lock_graph.c]: added
